modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -327,9 +331,15 @@ protected void onNodeLeft(UUID evtNodeId) {
     protected void enqueue(Collection<?> col) {
         assert Thread.holdsLock(this);
 
-        queue.add((Collection<R>)col);
-
-        cnt.addAndGet(col.size());
+        if(limitCnt <= 0 || limitCnt >= col.size()) {
+            queue.add((Collection<R>)col);
+            cnt.addAndGet(col.size());
+            limitCnt -= col.size();
+        } else {
+            int toAdd = limitCnt;
```

Comment:
limitCnt should be decreased in 'else' case as well

Reply:
Formally - yes, but this case will be the last pass under such condition and whole function will not be called any more.

Reply:
This is true for local query, but isn't for distributed
.
enqueue() method is called on every page received from remote node.

Assume we receive 2 pages from 2 nodes and try to enqueue them concurrently.
Both pages will be trimmed and added to queue, but the only first is expected to be added.

Reply:
I this case let us set 'limitCnt'=0 as soon as we have first node's portion that fulfills our limit. But arrived data from 2nd node still makes enqueue() to be called one more time. For 'limitCnt' =0 the first condition will fire as if it is no-limit case (limit<=0 is treated as no limit) what is not good specially here. So we need to add kind a 'limitReached' flag, set it in 'else' block and check at the method entry.

Reply:
I've made a PR with the only one additional field
https://github.com/apache/ignite/pull/6997/

Reply:
Concurrent test has been added